### ISSUE TITLE
[Bug]: Potential discrepancy between the RLT decoder and Eq. (2) of the paper

### ISSUE THREAD

**Issue by Ken1301225** [RLinf/RLinf#1391]
### Affected component

RLT token encoder-decoder implementation in `rlinf/models/embodiment/modules/rlt_token_transformer.py`, introduced by PR #1324.

### Bug description

I noticed a potential discrepancy between the RLT paper and the current decoder implementation.

Equation (2) of the paper defines an autoregressive reconstruction objective:

<img width="559" height="81" alt="Image" src="https://github.com/user-attachments/assets/78b4db6f-a486-4c08-ad04-e5202b9bbf7f" />

This suggests that the decoder is trained with shifted teacher-forcing inputs:

- Decoder input: `[z_rl, stopgrad(z_1), ..., stopgrad(z_{M-1})]`
- Reconstruction target: `[stopgrad(z_1), ..., stopgrad(z_M)]`
- Causal self-attention prevents position `i` from observing future embeddings

In the current implementation, `RLTSelfAttentionLayer` calls `nn.MultiheadAttention` without an `attn_mask` or `is_causal=True`. Furthermore, `RLTTokenDecoder` concatenates `rl_tokens` with learned `prefix_token_embed` queries instead of shifted ground-truth VLA embeddings.

Consequently, all reconstruction queries can attend to one another bidirectionally, and the sequence is reconstructed in parallel.

Could you clarify whether this is:

1. An intentional parallel query-based variant that performed better or was
   more efficient in your experiments; or
2. An implementation mismatch where the causal mask and teacher-forcing
   inputs are currently missing?

If this is intentional, documenting the difference from Eq. (2) and any relevant experimental observations would be helpful.

### Configuration YAML file

This appears to be configuration-independent whenever `openpi.use_rlt: true`.

The checked-in Stage 1 example uses:

```yaml
actor:
  model:
    openpi:
      use_rlt: true
      rlt_num_rl_tokens: 1
      rlt_num_layers: 2
      rlt_num_heads: 8
```

### Log file

Not applicable. This report concerns static attention semantics and decoder input construction rather than a runtime exception.

### Environment

Static code inspection on the current main branch after PR #1324.

Relevant implementation commit: `5769c6ebaa2c86e9ab37cc6ceb0d6b7168f67d70`.

The behavior follows PyTorch `nn.MultiheadAttention` semantics: without `attn_mask` or `is_causal=True`, attention is non-causal.

### Other reproduction info

Relevant locations:

- `RLTSelfAttentionLayer.forward`: the `self.self_attn(...)` call has only `key_padding_mask`, which does not impose causality.
- `RLTTokenDecoder.forward`: constructs `[rl_tokens, learned_prefix_tokens]`.
- The decoder calls each layer without any attention mask.
- Paper: https://arxiv.org/abs/2604.23073, Section III-A, Eq. (2).

I searched both open and closed issues in `RLinf/RLinf` for `RLT`, `RL Token`, `rlt_token_transformer`, `causal mask`, and `autoregressive decoder`, but did not find an existing issue covering this discrepancy.

### Before submitting a new issue...

- [x] Have you checked relevant issues, FAQs (https://rlinf.readthedocs.io/en/latest/rst_source/resources/faq.html), or asked the chatbot at the top right corner of the [documentation page](https://rlinf.readthedocs.io).

**Comment by zoeyuchao** [comment]
@TihayaKousaka , could u address this issue? thanks!

**Comment by tiny-xie** [comment]
Thank you for pointing this out. The subtle differences in our implementation actually stem from some considerations in the earliest version. We apologize that the final version was not fully aligned with the original paper. We will compare the performance of the two implementations as soon as possible and report the results back to you.
